src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,273 @@
+                    catch (OperationCanceledException)
+                    {
+                        _options.LogDebug(
+                            "Shutdown scheduled. Stopping by: {0}.",
+                            _options.ShutdownTimeout);
+
+                        shutdownTimeout.CancelAfterSafe(_options.ShutdownTimeout);
+
+                        shutdownRequested = true;
+                    }
+                }
+
+                // Work with the envelope while it's in the queue
+                foreach (var key in GetFlushableBuckets())
+                {
+                    // TODO: Check if a shutdown request has been made
+                    try
+                    {
+                        _options.LogDebug("Flushing metrics for bucket {0}", key);
+                        var bucket = Buckets[key];
+                        _captureMetrics(bucket.Values);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        _options.LogInfo("Shutdown token triggered. Time to exit.");
+                        return;
+                    }
+                    catch (Exception exception)
+                    {
+                        _options.LogError(exception, "Error while processing metric aggregates.");
+                    }
+                    finally
+                    {
+                        _options.LogDebug("Metric flushed for bucket {0}", key);
+                        Buckets.TryRemove(key, out _);
+                    }
+                }
+            }
+        }
+        catch (Exception e)
+        {
+            _options.LogFatal(e, "Exception in the Metric Aggregator.");
+            throw;
+        }
+    }
+
+    /// <summary>
+    /// Returns the keys for any buckets that are ready to be flushed (i.e. are for periods before the cutoff)
+    /// </summary>
+    /// <returns>
+    /// An enumerable containing the keys for any buckets that are ready to be flushed
+    /// </returns>
+    internal IEnumerable<long> GetFlushableBuckets()
+    {
+        var cutoff = MetricBucketHelper.GetCutoff();
+        foreach (var bucket in Buckets)
+        {
+            var bucketTime = DateTimeOffset.FromUnixTimeSeconds(bucket.Key);
+            if (bucketTime < cutoff)
+            {
+                yield return bucket.Key;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Stops the background worker and waits for it to empty the queue until 'shutdownTimeout' is reached
+    /// </summary>
+    /// <inheritdoc />
+    public void Dispose()
+    {
+        _options.LogDebug("Disposing MetricAggregator.");
+
+        if (_disposed)
+        {
+            _options.LogDebug("Already disposed MetricAggregator.");
+            return;
+        }
+
+        _disposed = true;
+
+        try
+        {
+            // Request the LoopTask stop.
+            _shutdownSource.Cancel();
```

Comment:
I don't see this being used anywhere. Is the top CancellationTokenSource supposed to be created out of composing with this one? (and if so, should we consider the options.ShutdownTimeout here too when calling cancel?